src/Dialogs/Cloud/WeGlideUploadDialog.cpp

```diff
@@ -0,0 +1,191 @@
+      Json::Lookup(user, "name")->get_string().c_str());
+
+    auto aircraft = Json::Lookup(json_value, "aircraft")->as_object();
+
+    response.glider.id = Json::Lookup(aircraft, "id")->get_int64();
+    response.glider.name = UTF8ToWideConverter(Json::Lookup(aircraft, "name")->get_string().c_str());
+    response.glider.kind = UTF8ToWideConverter(Json::Lookup(aircraft, "kind")->get_string().c_str());
+    response.glider.sc_class = UTF8ToWideConverter(Json::Lookup(aircraft, "sc_class")->get_string().c_str());
+
+  }
+
+  bool UploadFlightDialog(Path igc_path, WeGlide::Pilot pilot, uint_least32_t glider_id) {
+    const WeGlideSettings& weglide_settings =
+      CommonInterface::GetComputerSettings().weglide;
+    StaticString<0x1000> msg;
+    NarrowString<0x200> url(weglide_settings.default_url);
+    url += "/igcfile";
+
+    try {
+      if (glider_id == 0)
+        glider_id = CommonInterface::GetComputerSettings().plane
+        .weglide_glider_type;
+      if (pilot.id == 0) {
+        pilot = weglide_settings.pilot;
+      }
+      UploadResponse response;
+      PluggableOperationEnvironment env;
+      LogFormat(_("WeGlide Upload: %s"), igc_path.c_str());
+      if (!File::Exists(LocalPath(igc_path))) {
+        msg.Format(_("Error! File '%s' doesn't exist!"), igc_path.c_str());
+        LogFormat(_("WeGlide Upload: %s"), msg.c_str());
+        ShowMessageBox(msg, _T("WeGlide Upload"), MB_OK | MB_ICONEXCLAMATION);
+
+        return false;
+      }
+
+      if (ShowCoDialog(UIGlobals::GetMainWindow(), UIGlobals::GetDialogLook(),
+        _("Download"), UpdateTask(igc_path, url, pilot,
+          glider_id, response, env), &env) == false)
+      {
+        return false;
+      }
+
+      msg.Format(_("File upload '%s' was succesfull!\n\n"), igc_path.c_str());
+      msg.AppendFormat(_("%-20s: %u\n"), _("flight_id"), response.flight_id);
+      // Unfortunately on Linux you cannot use the NarrowString itself, you have
+      // to convert this with '.c_str()'
+      msg.AppendFormat(_("%-20s: %s\n"), _("scoring_date"),
+        response.scoring_date.c_str());
+      msg.AppendFormat(_("%-20s: %s (%d)\n"), _("pilot"),
+        response.pilot.name.c_str(), response.pilot.id);
+      msg.AppendFormat(_("%-20s: %s (%u)\n"), _("aircraft type"),
+        response.glider.name.c_str(), response.glider.id);
+      msg.AppendFormat(_("%-20s: %s: %s "), _("glider"), _("reg"),
+        response.registration.c_str());
+      msg.AppendFormat(_("%s: %s\n"), _("cid"),
+        response.competition_id.c_str());
+
+      ShowMessageBox(msg, _T("WeGlide Upload"), MB_OK | MB_ICONEXCLAMATION);
+
+      return true;  // upload succesful!
+    }
+    catch (const std::exception& e) {
+      // Check for errors
+      const char* curl_response[] = {
+        "CURL failed: The requested URL returned error: 406",
+        "CURL failed: The requested URL returned error: 400",
+        "CURL failed: Failed to open/read local data from file/application"
```

Comment:
Bad hard-coded error messages. You shouldn't ever use strcmp() to check for error types.

Reply:
Why not? The user can see every time a changed error message as it is, and so every change in the system (curl) don't make a problem - he get the new error message. Only in this 3 cases the xcsoar user get an additional hint, why this message occurs - and this can help a lot to understand, what is going wrong, at least at the error code 406.

Reply:
What if the next CURL version changes the wording of these error messages? This will break your code. Your code is fragile and error prone. If you want to check for certain error conditions, never check human-readable strings, always check error/status codes instead.

Reply:
Yes, I know - but after this CURL change it happens nothing wrong: The changed error text will be showed, nothing else.
But ok, how can I get the error code from the curl component directly? Unfortunately I couldn't found  any hint ;-(

Reply:
https://curl.se/libcurl/c/CURLINFO_RESPONSE_CODE.html

Reply:
Ok, I will try it over our curl component!

Reply:
Unfortunately it's not as easy as it seems ;-(
In the event of an error - regardless of whether the error comes from the Curl component or the HTTP response shows an error, an exception is thrown by the Easy component in Adapter.cxx - but in the catch branch I have no chance of going back to the Easy Componente (passed it to Curl::CoStreamRequest beforehand with std::move()) - and so I have no chance to get the error code there either...
Now I see several possible solutions again, but all of them would change the code in the CurlResponseHandlerAdapter component (Adapter.cxx) - of course then with possible side effects on other accesses :( E.g. you could do without the exception in CurlResponseHandlerAdapter::Done - and set the response.status to the error code, or you 'build' a predefined exception string with the error codes, which you then parse (e.g. FormatRuntimeError("CURL failed: %d, %s", result, error_buffer) or . possibly also a json string? ) - I had also checked other possibilities, but then all discarded...
Maybe you have a better and simpler approach, but I can't get any further here...

Reply:
You can disable `CURLOPT_FAILONERROR` if you are interested in handling non-success status codes.

Reply:
I know, than the result is CURLE_OK - but only inside the Easy/Adapter object...
My favorite is the JSON string as an exception message instead of plain text, like:
```
throw FormatRuntimeError("{'id':'CURL failed','curl_error': %d, 'http_error': %d,"
        " 'message': '%s'}", result, http_result, msg);
```
in function CurlResponseHandlerAdapter::Done(CURLcode result) (Adapter.cxx)

Reply:
> I know, than the result is CURLE_OK

Of course, and from there, you check the status code.

Reply:
Where? This isn't possible inside the UploadFlight.cpp, because the easy object is gone - and 0!

Reply:
Awaiting the `CoStreamRequest` returns a `CoResponse` which has `status` field.
https://github.com/XCSoar/XCSoar/blob/faa9f790bd188ed3a02d108b33a53bb884a5cf65/src/Cloud/weglide/UploadFlight.cpp#L79-L80
Here it is - `response`.

Reply:
Of course, but here I cannot call `easy.GetInfo(CURLINFO_RESPONSE_CODE, &error_code);` because easy is 0, or 
`curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &error_code);`  because (Global) curl responses this funktion likewise with 0!
I had checked it very, very often...
Maybe there is a 'living' object, which gives the right answer, but I don't know anything!

As I said: Up to know my only choice is the changing of the exception text (with  `easy.SetFailOnError();`) -or a bigger change in Adapter.cxx (but this makes much more trouble)

Reply:
Why do you want to query `CURLINFO_RESPONSE_CODE` when you already have the same piece of information in `response.status`?
I don't understand your problem.